lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +194,246 @@ class Loader {
-      process.emitWarning(
-        'The dynamicInstantiate loader hook has been removed.');
-    }
-    if (getFormat !== undefined) {
-      this._getFormat = FunctionPrototypeBind(getFormat, null);
-    }
-    if (getSource !== undefined) {
-      this._getSource = FunctionPrototypeBind(getSource, null);
-    }
-    if (transformSource !== undefined) {
-      this._transformSource = FunctionPrototypeBind(transformSource, null);
-    }
-    if (getGlobalPreloadCode !== undefined) {
-      this._getGlobalPreloadCode =
-        FunctionPrototypeBind(getGlobalPreloadCode, null);
+    const namespaces = new Array(specifiers.length);
+
+    for (const { 0: i, 1: specifier } of specifiers.entries()) {
+      const job = await this.getModuleJob(specifier, parentURL);
+      const { module } = await job.run();
+
+      namespaces[i] = module.getNamespace();
     }
+
+    return wasArr ?
+      namespaces :
+      namespaces[0];
   }
 
-  runGlobalPreloadCode() {
-    if (!this._getGlobalPreloadCode) {
-      return;
+  /**
+   * Provide the ready-to-use source (output is something node understands).
+   *
+   * The internals of this WILL change when chaining is implemented,
+   * depending on the resolution/consensus from #36954
+   * @param {string} url The URL/path of the module to be loaded
+   * @param {Object} context Metadata about the module
+   * @returns {Object}
+   */
+  async load(url, context = {}) {
+    const { 0: defaultLoader, 1: customLoader } = this.#loaders;
+
+    const loader = customLoader || defaultLoader;
+    const loaded = await loader(url, context, defaultLoader);
+
+    if (typeof loaded !== 'object') {
+      throw new ERR_INVALID_RETURN_VALUE(
+        'object',
+        'loader load',
+        loaded,
+      );
     }
-    const preloadCode = this._getGlobalPreloadCode();
-    if (preloadCode === null) {
-      return;
+
+    const {
+      format,
+      source,
+    } = loaded;
+
+    if (format === null) {
+      const dataUrl = RegExpPrototypeExec(
+        /^data:([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/,
+        url,
+      );
+
+      throw new ERR_INVALID_MODULE_SPECIFIER(
+        url,
+        dataUrl ? `has an unsupported MIME type "${dataUrl[1]}"` : ''
+      );
     }
 
-    if (typeof preloadCode !== 'string') {
-      throw new ERR_INVALID_RETURN_VALUE(
-        'string', 'loader getGlobalPreloadCode', preloadCode);
+    if (typeof format !== 'string') {
+      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
+        'string',
+        'loader resolve',
+        'format',
+        format,
+      );
     }
-    const { compileFunction } = require('vm');
-    const preloadInit = compileFunction(preloadCode, ['getBuiltin'], {
-      filename: '<preload>',
-    });
-    const { NativeModule } = require('internal/bootstrap/loaders');
-
-    preloadInit.call(globalThis, (builtinName) => {
-      if (NativeModule.canBeRequiredByUsers(builtinName)) {
-        return require(builtinName);
+
+    if (
+      source &&
```

Comment:
I think this new case cannot happen: the first check is a truthy check, which assures `source` is not falsy (and `null` is falsy).

Reply:
My suggestion is to remove the truthy check and use `source != null` instead, which would matter only for the empty string. It's a suggestion that only makes sense if we go for the nullish way instead of falsy in the rest of the implementation

Reply:
gah, sorry! I was looking at this on my mobile earlier (which is also why I didn't see you had left comments citing the specific examples I subsequently requested).